src/Middleware/HttpLogging/src/W3CLogger.cs

```diff
@@ -0,0 +1,130 @@
+                    case nameof(DateTime):
+                        DateTime dt = DateTime.Parse(val, CultureInfo.InvariantCulture);
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Date)] = dt.ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Time)] = dt.ToString("HH:mm:ss", CultureInfo.InvariantCulture);
+                        // We estimate time elapsed by diffing the current time with the time at which the middleware processed the request/response.
+                        // This will represent the time in whole & fractional milliseconds.
+                        var elapsed = DateTime.Now.Subtract(dt);
+                        elements[BitOperations.Log2((int)W3CLoggingFields.TimeTaken)] = elapsed.TotalMilliseconds.ToString(CultureInfo.InvariantCulture);
+                        break;
+                    case nameof(HeaderNames.Server):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerName)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.RemoteIpAddress):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ClientIpAddress)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.LocalIpAddress):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerIpAddress)] = val.Trim();
+                        break;
+                    case nameof(ConnectionInfo.LocalPort):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.ServerPort)] = val.Trim();
+                        break;
+                    case nameof(HttpContext.User):
+                        elements[BitOperations.Log2((int)W3CLoggingFields.UserName)] = val.Trim();
+                        break;
+                    case nameof(HeaderNames.Cookie):
+                        // Cookie can have whitespace - we replace whitespace characters with the '+' character
+                        elements[BitOperations.Log2((int)W3CLoggingFields.Cookie)] = Regex.Replace(val.Trim(), @"\s", "+");
+                        break;
+                    default:
+                        break;
+                }
+            }
+            StringBuilder sb = new StringBuilder();
+            for (int i = 0; i < elements.Length; i++)
+            {
+                if (_loggingFields.HasFlag((W3CLoggingFields)(1 << i)))
+                {
+                    // If the element was not logged, or was the empty string, we log it as a dash
+                    if (String.IsNullOrEmpty(elements[i]))
+                    {
+                        sb.Append("- ");
+                    }
+                    else
+                    {
+                        sb.Append(elements[i] + " ");
+                    }
+                }
+            }
+            return sb.ToString().Trim();
```

Comment:
Trim will allocate a new string. Potentially this could be quite a large string. Instead of trimming at the end, update the logic to only add separator space (I'm guessing that is what this is for) if it is required.